android/sdk/src/main/java/com/taobao/weex/bridge/WXBridgeManager.java

```diff
@@ -2948,7 +2951,69 @@ public int callUpdateAttrs(String instanceId, String ref, HashMap<String, String
 
     return IWXBridge.INSTANCE_RENDERING;
   }
+  private boolean shouldReportGPULimit() {
+    IWXConfigAdapter adapter = WXSDKManager.getInstance().getWxConfigAdapter();
+    boolean report_gpu_limited_layout = false;
+    float sample_rate_of_report = 0;
+    if (adapter != null) {
+      try {
+        sample_rate_of_report = Float.parseFloat(adapter
+                .getConfig("android_weex_test_gpu",
+                        "sample_rate_of_report",
+                        "0"));
+      }catch(Exception e){
+        WXLogUtils.e(WXLogUtils.getStackTrace(e));
+      }
+      WXLogUtils.i("sample_rate_of_report : " + sample_rate_of_report);
+      if(Math.random() < sample_rate_of_report){
+        report_gpu_limited_layout = true;
+      }
+    }
+    return report_gpu_limited_layout;
+  }
 
+  private void reportIfReachGPULimit(String instanceId,String ref,GraphicSize layoutSize){
+    float limit = WXRenderManager.getOpenGLRenderLimitValue();
+    if(limit > 0 && (layoutSize.getHeight() > limit || layoutSize.getWidth() > limit)){
+      Map<String, String> ext = new ArrayMap<>();
+      WXComponent component = WXSDKManager.getInstance().getWXRenderManager().getWXComponent(instanceId,ref);
+      ext.put("GPU limit",String.valueOf(limit));
+      ext.put("component.width",String.valueOf(layoutSize.getWidth()));
+      ext.put("component.height",String.valueOf(layoutSize.getHeight()));
+      if (component.getComponentType() != null && !component.getComponentType().isEmpty()) {
+        ext.put("component.type", component.getComponentType());
+      }
+      if (component.getStyles() != null && !component.getStyles().isEmpty()) {
+        ext.put("component.style", component.getStyles().toString());
+      }
+      if (component.getAttrs() != null && !component.getAttrs().isEmpty()) {
+        ext.put("component.attr", component.getAttrs().toString());
+      }
+      if (component.getEvents() != null && !component.getEvents().isEmpty()) {
+        ext.put("component.event", component.getEvents().toString());
+      }
+      if (component.getMargin() != null) {
+        ext.put("component.margin", component.getMargin().toString());
+      }
+      if (component.getPadding() != null) {
+        ext.put("component.padding", component.getPadding().toString());
+      }
+      if (component.getBorder() != null) {
+        ext.put("component.border", component.getBorder().toString());
+      }
+      Map<String,Object> map = new JSONObject();
+      map.putAll(ext);
+      WXSDKManager.getInstance().getSDKInstance(instanceId).setComponentsInfoExceedGPULimit((JSONObject)map);
```

Comment:
Just `getComponentsExceedGPULimit(map)` is enough, no need to upcast `map` to `JSONObject`